### PR TITLE
Referenced Azure.Identity from Azure.Messaging.EventHubs.Samples

### PR DESCRIPTION
**Pull request by albertodenatale**
Hi
The change aims to fix this building error occurring when referencing ```Azure.Identity```  from ```Azure.Messaging.EventHubs.Samples```.

"_Detected package downgrade: System.Runtime.CompilerServices.Unsafe from 4.6.0 to 4.5.2._"

Azure.Identity -> System.Text.Json 4.6.0 -> System.Runtime.CompilerServices.Unsafe (>= 4.6.0) 
Azure.Messaging.EventHubs.Tests -> System.Runtime.CompilerServices.Unsafe (>= 4.5.2)	

It upgrades the library ```System.Runtime.CompilerServices.Unsafe``` globally for the sdk, in "Track 2 specific versions" section.

Annotation 2019-10-24 074815

### CODE REVIEW

**Review by jsquire — CHANGES_REQUESTED** [review]
Hi Alberto.   Thank you for your continued efforts to improve the Event Hubs client, and for putting this PR together!

Since `Identity` is an external dependency, I don't think we want to take a hard project reference here.   Rather, we'll need to use an abstraction that allows toggling between the project or package references depending on the repository configuration.  The good news is that we needed to do the same abstraction for the Event Hubs Checkpoint Store package, so there's a full example to use as a point of reference.

I'd recommend taking a look at the root of the [Azure.Messaging.CheckpointStore.Blobs](https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/eventhub/Azure.Messaging.EventHubs.CheckpointStore.Blobs) project and take note of the file [Azure.Client.Library.References.props](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/eventhub/Azure.Messaging.EventHubs.CheckpointStore.Blobs/Azure.Client.Library.References.props).  

In there, we have conditional references for the dependencies which are external to the `CheckpointStore.Blobs` package itself but within the Azure SDK repository.    We should do the same thing with the `Identity` library here.    For usage, the [Azure.Messaging.EventHubs.CheckpointStore.Blobs.Samples.csproj](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/eventhub/Azure.Messaging.EventHubs.CheckpointStore.Blobs/samples/Azure.Messaging.EventHubs.CheckpointStore.Blobs.Samples.csproj) is probably the simplest illustration,  starting on line 18.

If you've got questions or would like to chat further, feel free to drop a comment and reference me.  I'm happy to help as needed.

**Comment by jsquire** [comment]
@pakrym:  Since you've done a lot of the version management across the System references for the track two libraries,  do you mind sharing your thoughts on the potential upgrade of `System.Runtime.CompilerServices.Unsafe` to `4.6.0`?

Alternatively, I *think* it would be safe to remove from `Event Hubs` and its tests.  A quick check of doing so doesn't seem to have caused issue, but this was part of of pre-release package set that we needed for some of the pre-.NET Core 3 functionality.   Any chance that you have any better insight than I do on whether we can trim that out?

**Comment by pakrym** [comment]
@jsquire yeah, just drop the dependency from event hubs. If it compiles everything is fine :)

**Comment by albertodenatale** [comment]
@jsquire thank you very much for you feedback, I am very excited for the opportunity of coding with a fine selection of professionals such as you all are.

I will go in the direction you kindly suggested and come back to you.

**Comment by albertodenatale** [comment]
@jsquire I wonder if it would make sense to remove ```System.Runtime.CompilerServices.Unsafe``` completely? 

As far as I can see, is currently referenced only from ```Azure.Messaging.EventHubs.CheckpointStore.Blobs.csproj``` and from ```Azure.Messaging.EventHubs.CheckpointStore.Blobs.Tests```.

Did the changes we discussed, very nice solution the conditional references. Look forward to having your feedback.

**Comment by jsquire** [comment]
> @jsquire I wonder if it would make sense to remove `System.Runtime.CompilerServices.Unsafe` completely?
> 
> As far as I can see, is currently referenced only from `Azure.Messaging.EventHubs.CheckpointStore.Blobs.csproj` and from `Azure.Messaging.EventHubs.CheckpointStore.Blobs.Tests`. Tried to remove them from there and it actually builds.

Excellent catch!   The [dependency report](https://azuresdkartifacts.blob.core.windows.net/azure-sdk-for-net/dependencies/dependencies.html) would seem to agree with you.   I think it's a great idea to try.   CI and/or the associated test suites should surface any problems if it is used elsewhere.

**Review by jsquire — COMMENTED** [review]
Changes look good to me.   Not sure why the CI build is complaining about a downgrade.   I wonder if there's an implicit dependency on `Unsafe`

**Comment by pakrym** [comment]
You need to remove `Azure.Messaging.EventHubs.CheckpointStore.Blobs.Tests -> System.Runtime.CompilerServices.Unsafe ` reference.

**Comment by jsquire** [comment]
> You need to remove `Azure.Messaging.EventHubs.CheckpointStore.Blobs.Tests -> System.Runtime.CompilerServices.Unsafe ` reference.

Good catch.   Forgot about those.

File: sdk/eventhub/Azure.Messaging.EventHubs/Azure.Client.Library.References.props [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<Project ToolsVersion="Current" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/Azure.Client.Library.References.props]: One last nit, and apologies for not having thought of this sooner.    Could we move this file to the `samples` directory, as it is currently only applicable to the samples.

By contrast, the one in the Blobs CheckpointStore library is used across the project as a common point of reference.

**Review by jsquire — COMMENTED** [review]
Thanks for continuing to iterate on this, @albertodenatale.    Looks like the CI runs are happy now.  I left one additional small change request, if you would be so kind.   Then, I think we're good to merge.

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Azure.Messaging.EventHubs.Tests.csproj [hunk 2]
```diff
@@ -35,4 +34,7 @@
 
   <!-- Import the Azure.Core test framework (shared source) -->
   <Import Project="..\..\..\core\Azure.Core\tests\TestFramework.props" />
+  
+  <!-- Import the references to the Azure client libraries -->
+  <Import Project="$(MSBuildThisFileDirectory)..\Azure.Client.Library.References.props" />
```
Comment (albertodenatale) [inline sdk/eventhub/Azure.Messaging.EventHubs/tests/Azure.Messaging.EventHubs.Tests.csproj]: > One last nit, and apologies for not having thought of this sooner. Could we move this file to the `samples` directory.

@jsquire 

Thank you for your nit,

Shall I update this into ```$(MSBuildThisFileDirectory)..\samples\Azure.Client.Library.References.props```

Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/tests/Azure.Messaging.EventHubs.Tests.csproj]: I think I need to mea cupla here; I overlooked that this was used in both the `samples` and `tests`.    I withdraw my nit.    Your current organization is the correct one, given multiple references.   Sorry about that.

File: sdk/eventhub/Azure.Messaging.EventHubs/Azure.Client.Library.References.props (hunk 1 above)
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/Azure.Client.Library.References.props]: This is invalid;  it was pointed out to me that references to this are made by both tests and samples.   It's in the correct location.   My bad.

**Review by jsquire — APPROVED** [review]
Changes look good to me.   Going to run CI one last time and, assuming all goes well, will merge at that point.

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Azure.Messaging.EventHubs.Tests.csproj (hunk 2 above)
Comment (albertodenatale) [reply sdk/eventhub/Azure.Messaging.EventHubs/tests/Azure.Messaging.EventHubs.Tests.csproj]: No worries, it was a good one 👊

**Pull request merged**